File: src/main/java/com/youngjun/auth/core/api/support/response/AuthResponse.java
package com.youngjun.auth.core.api.support.response;

import com.youngjun.auth.core.api.support.error.AuthErrorMessage;
import com.youngjun.auth.core.api.support.error.AuthErrorType;

public record AuthResponse<T>(AuthStatusType status, T data, AuthErrorMessage error) {
	public static AuthResponse<Void> success() {
		return new AuthResponse<>(AuthStatusType.SUCCESS, null, null);
	}

	public static <T> AuthResponse<T> success(T data) {
		return new AuthResponse<>(AuthStatusType.SUCCESS, data, null);
	}

	public static AuthResponse<Void> error(AuthErrorType authErrorType) {
		return new AuthResponse<>(AuthStatusType.ERROR, null, new AuthErrorMessage(authErrorType));
	}

	public static AuthResponse<Void> error(AuthErrorType authErrorType, Object data) {
		return new AuthResponse<>(AuthStatusType.ERROR, null, new AuthErrorMessage(authErrorType, data));
	}
}

File: src/main/kotlin/com/youngjun/auth/core/api/support/response/AuthResponse.kt
package com.youngjun.auth.core.api.support.response

import com.youngjun.auth.core.api.support.error.ErrorMessage
import com.youngjun.auth.core.api.support.error.ErrorType

data class AuthResponse<T> private constructor(
    val status: ResultType,
    val data: T? = null,
    val error: ErrorMessage? = null,
) {
    companion object {
        fun success(): AuthResponse<Any> = AuthResponse(ResultType.SUCCESS, null, null)

        fun <S> success(data: S): AuthResponse<S> = AuthResponse(ResultType.SUCCESS, data, null)

        fun error(
            error: ErrorType,
            errorData: Any? = null,
        ): AuthResponse<Any> = AuthResponse(ResultType.ERROR, null, ErrorMessage(error, errorData))
    }
}
